module/preference/PreferenceDialog.kt

```diff
@@ -0,0 +1,335 @@
+
+    sealed interface Preference<T> {
+        var value: T
+        var label: String
+
+        @Composable fun display()
+        fun valid(): Boolean
+
+        class TextInput(initial: String, override var label: String, private var placeholder: String,
+                        var validator: (String) -> Boolean = { true }) : Preference<String> {
+
+            override var value by mutableStateOf(initial)
+
+            override fun valid(): Boolean {
+                return validator(value)
+            }
+
+            @Composable
+            override fun display() {
+                var border = Form.Border(1.dp, RoundedCornerShape(Theme.ROUNDED_CORNER_RADIUS)) {
+                    if (this.valid()) {
+                        Theme.studio.border
+                    } else {
+                        Theme.studio.errorStroke
+                    }
+                }
+
+                Field(label = label) {
+                    TextInput(
+                        value = value,
+                        placeholder = placeholder,
+                        border = border,
+                        onValueChange = { value = it }
+                    )
+                }
+            }
+        }
+
+        class Checkbox(initial: Boolean, override var label: String): Preference<Boolean> {
+            override var value by mutableStateOf(initial)
+            @Composable
+            override fun display() {
+                Field(label = label) {
+                    Checkbox(
+                        value = value,
+                        onChange = { value = it }
+                    )
+                }
+            }
+
+            override fun valid(): Boolean {
+                return true
+            }
+        }
+    }
+
+
+    class PreferencesForm : State {
+        // Graph Visualiser Preferences
+        var graphOutput = Preference.Checkbox(appData.graphOutput, Label.ENABLE_GRAPH_OUTPUT)
+        // Project Manager Preferences
+        val ignoredPathsData = appData.ignoredPaths.toString()
+        var ignoredPaths = Preference.TextInput(ignoredPathsData.substring(1, ignoredPathsData.length - 1),
+            Label.PROJECT_IGNORED_PATHS, IGNORED_PATHS_PLACEHOLDER)
+        // Query Runner Preferences
+        var limit = Preference.TextInput(appData.limit, Label.SET_QUERY_LIMIT, QUERY_LIMIT_PLACEHOLDER) { it.toLongOrNull() != null }
+        // Text Editor Preferences
+        var autoSave = Preference.Checkbox(appData.autoSave, Label.ENABLE_EDITOR_AUTOSAVE)
+
+        override fun cancel() {
+            StudioState.preference.openPreferenceDialog.close()
+        }
+
+        fun apply() {
+            trySubmit()
+        }
+
+        fun ok() {
+            apply()
+            cancel()
+        }
+
+        override fun isValid(): Boolean {
+            return graphOutput.valid() && ignoredPaths.valid() && limit.valid() && autoSave.valid()
+        }
```

Comment:
As we add preferences, we may want to collect all of the preferences into a list and validate them all. We'll also want to use this list for other purposes. Maybe a tree of preferences would be more correct, and we could traverse this tree to verify every leaf is valid (in which case isValid would return true).